### PR TITLE
WIP: Use KVO to install cluster

### PR DESCRIPTION
**Pull request by derekparker**
This patch installs core Kubernetes components using the KVO. 

This mode in the KVO is triggered by specifying a "DesiredVersion" without a "CurrentVersion".

### CODE REVIEW

**Comment by derekparker** [comment]
Going to need KVO to create the AppVersion kind on startup for this to work.

**Comment by derekparker** [comment]
Not quite ready for review, need some more coordination with KVO. KVO must create the app version from this new cluster configure object which now holds the initial version.

**Comment by derekparker** [comment]
Alright, so this is good enough for initial review. It won't work exactly because I need to build a new dev KVO image after https://github.com/coreos-inc/kube-version-operator/pull/316 merges, however it's ready for at least the Terraform changes to be reviewed.

Just to explain this new bootstrap method:

All control plane manifests have been moved out of this repo and into the KVO upgrade spec. When the master node starts, it will run KVO via container image in "render" mode, which will render all assets, including bootstrap assets, onto disk using the cluster-config for templating. From there, the rest of the process continues as normal. Bootkube will start and kick the rest off.

**Comment by aaronlevy** [comment]
@sym3tri @s-urbaniak @alexsomesan -- could you take a review pass at this (mostly for terraform stuff). It's getting pretty close - and to keep the timezone delays to a minimum, hoping to start getting some feedback.

**Comment by derekparker** [comment]
With this last WIP commit the AWS smoke tests are passing when run on my machine (the CI test seems to be failing immediately with no actual output as to why... most likely because I need to rebase and pull in other changes to make Jenkins happy).

I can rebase first thing tomorrow, however I would really appreciate some feedback on this new approach. I've outlined the new bootstrap method in a comment above, however I can expand on that if need be. I know that I will need to port my Terraform changes to other platforms, as they are all AWS specific at the moment, but initial feedback on the AWS changes and overall direction would be greatly appreciated.

cc @sym3tri @s-urbaniak @alexsomesan

File: modules/bootkube/assets.tf [hunk 1]
```diff
@@ -178,17 +149,68 @@ resource "local_file" "bootkube_sh" {
...
+    advertise_address   = "${var.advertise_address}"
+    etcd_servers = "${
+      var.experimental_enabled
+        ? format("https://%s:2379", cidrhost(var.service_cidr, 15))
+        : var.etcd_ca_cert_pem == ""
+          ? join(",", formatlist("http://%s:2379", var.etcd_endpoints))
+          : join(",", formatlist("https://%s:2379", var.etcd_endpoints))
+      }"
+
+    master_count = "${var.master_count}"
```
Comment (s-urbaniak) [inline modules/bootkube/assets.tf]: let's group the single line vars and sort them alphabetically and only let the `etcd_server` variable have a newline between them like so:
```
vars {
   advertise_address      = "${var.advertise_address}"
   cloud_config_path      = "${var.cloud_config_path}"
   cloud_provider_profile = "${var.cloud_provider != "" ? "${var.cloud_provider}" : "metal"}"
   cluster_cidr           = "${var.cluster_cidr}"
...
   etcd_servers = "${
      var.experimental_enabled
        ? format("https://%s:2379", cidrhost(var.service_cidr, 15))
        : var.etcd_ca_cert_pem == ""
          ? join(",", formatlist("http://%s:2379", var.etcd_endpoints))
          : join(",", formatlist("https://%s:2379", var.etcd_endpoints))
      }"
  }
}
```

We found this is easier for diff'ing than with logical grouping.

File: tests/smoke/aws/vars/aws-ca.tfvars.json [hunk 2]
```diff
@@ -7,7 +7,7 @@
 
     "tectonic_etcd_servers": [""],
 
-    "tectonic_base_domain": "tectonic.dev.coreos.systems",
+    "tectonic_base_domain": "dparker.k8s.work",
```
[the remark this answers is not in the document]
Comment (derekparker) [reply tests/smoke/aws/vars/aws-ca.tfvars.json]: Heh, yes. Had to modify for local testing.

File: modules/bootkube/resources/bootkube.service [hunk 3]
```diff
@@ -1,8 +1,8 @@
 [Unit]
 Description=Bootstrap a Kubernetes cluster
 ConditionPathExists=!/opt/tectonic/init_bootkube.done
-Wants=kubelet.service
-After=kubelet.service
+Wants=kvo.service
+After=kvo.service
```
[the remark this answers is not in the document]
Comment (derekparker) [reply modules/bootkube/resources/bootkube.service]: Done

File: Documentation/dev/node-bootstrap-flow.md [hunk 4]
```diff
@@ -20,7 +20,9 @@ When a cluster node is being bootstrapped from scratch, it goes through several
 
 Additionally, only on one of the master nodes the following kubernetes bootstrapping happens:
 
-1. `bootkube.service` is started after `kubelet.service` start
+1. `kvo.service` is started after `kubelet.service` start
+1. `kvo.service` renders all bootstrap and self-hosted control plane assets
```
Comment (aaronlevy) [inline Documentation/dev/node-bootstrap-flow.md]: Wondering if we should just squash the kvo.service step as part of bootkube.service? Really it's "bootstrap" service - and it should only ever run once on a single node anyway (I believe as this stands kvo.service would run once on every master)

**Comment by derekparker** [comment]
Looks like the last CI run failed because of a Jenkins issue... can someone kick it over? @s-urbaniak

**Comment by derekparker** [comment]
Since I'll no longer be maintaining this PR, here's a few notes for whoever comes next:

* The KVO render step has been moved to its own binary (https://github.com/coreos-inc/kube-version-operator/tree/master/cmd/render-manifests)
* A new image & repo has been created for it: https://quay.io/repository/coreos/kube-version-renderer
* The KVO service needs to be changed in order to use this new image, instead of using the actual operator
* We no longer need to put docker creds in place, as the renderer image is not private
* Most likely want to put the render & bootkube step into a single bootstrap service file
* The only platform fully updated for this new flow is AWS, so other platforms will have to be finished as well
* Must ensure that vanilla k8s continues to work (the discussion happened ~3 weeks ago so I forget everything that is involved in this)

I think that is it. I think @diegs is most likely the next person to pick this up, or at least the point of contact until somebody else does.

**Comment by diegs** [comment]
Thanks @derekparker for all your hard work. For posterity this is on ice for the time being and will be picked up for the January release at the earliest.

**Comment by diegs** [comment]
Resolved in #2702

**Pull request closed without merging**